### DISCUSSION TITLE
Can Langfuse be used to record the records of ConversationChain in LangChain?

### DISCUSSION THREAD

**Discussion by lyzltysgithub** [langfuse/langfuse#1225] [Support] [answered]
Hello! I have seen in the tutorial (https://langfuse.com/docs/integrations/langchain/example-python) that it mentioned the use of langfuse with Sequential Chain and RetrievalQA in LangChain, which can be invoked using overall_chain.run("Tragedy at sunset on the beach", callbacks=[langfuse_handler]). However, I am unable to invoke langfuse using ConversationChain (from langchain.chains import ConversationChain).

Specifically, in the code:

def generate_sql_query(data_query,history,owner_id):
    llm_adaptor = LLMAdaptor(model='gpt-4-turbo',owner_id=owner_id)
    memory = ConversationBufferWindowMemory(k=3) 
    for input_message, output_message in history:
        memory.save_context({"input": input_message}, {"output": output_message})
    conversation = ConversationChain(llm=llm_adaptor,memory=memory,verbose=False)
    # result = conversation.predict(input=data_query)
    result = conversation.invoke({"input":data_query},config={"callbacks":[langfuse_handler]})
    return result

I encountered the following error:

  File "/tmp/ipykernel_44533/415170381.py", line 127, in generate_sql_query
    result = conversation.invoke({"input":data_query},config={"callbacks":[langfuse_handler]})
AttributeError: 'ConversationChain' object has no attribute 'invoke'

I also tried result = conversation.run(input=data_query, config={"callbacks": [langfuse_handler]}), but it didn’t work either.

Am I writing the code incorrectly? Do you have any suggestions? Or does ConversationChain not support langfuse?

Thank you very much!

**Comment by marcklingen** [comment, marked as answer]
I've just added it to the cookbook, hope this helps!
https://langfuse.com/docs/integrations/langchain/example-python#conversationchain

```python
from langfuse.callback import CallbackHandler
from langchain.chains import ConversationChain
from langchain.memory import ConversationBufferMemory
from langchain_openai import OpenAI
 
llm = OpenAI(temperature=0)
conversation = ConversationChain(
    llm=llm, memory=ConversationBufferMemory()
)
# Create a callback handler with a session
langfuse_handler = CallbackHandler(session_id="conversation_chain")

conversation.predict(input="Hi there!", callbacks=[langfuse_handler])
conversation.predict(input="How to build great developer tools?", callbacks=[langfuse_handler])
conversation.predict(input="Summarize your last response", callbacks=[langfuse_handler])
```

**Reply by lyzltysgithub** [reply]
Thank you for adding it to the cookbook! It's really helpful.
